tutorial/LFRic/background/LFRic_kernel.md

```diff
@@ -0,0 +1,241 @@
+object. As a rule, object data are `private` as illustrated in the
+`setval_field_w0_kernel_type` definition above. Object procedures can be
+`private` (if used only by the parent object) or `public`ly available
+(e.g. `subroutine setval_field_w0_code` is called in the generated PSy layer).
+
+### Kernel subroutine
+
+#### Argument list and declarations
+
+Running PSyclone
+[kernel stub generator](https://psyclone.readthedocs.io/en/stable/stub_gen.html)
+on `setval_field_w0_kernel_mod.f90` produces the argument list and
+declarations for the `setval_field_w0_code` enclosed in a module:
+
+```fortran
+   MODULE setval_field_w0_mod
+    IMPLICIT NONE
+    CONTAINS
+    SUBROUTINE setval_field_w0_code(nlayers, field_1_w0, rscalar_2, ndf_w0, undf_w0, map_w0)
+      USE constants_mod, ONLY: r_def, i_def
+      IMPLICIT NONE
+      INTEGER(KIND=i_def), intent(in) :: nlayers
+      INTEGER(KIND=i_def), intent(in) :: ndf_w0
+      INTEGER(KIND=i_def), intent(in), dimension(ndf_w0) :: map_w0
+      INTEGER(KIND=i_def), intent(in) :: undf_w0
+      REAL(KIND=r_def), intent(in) :: rscalar_2
+      REAL(KIND=r_def), intent(inout), dimension(undf_w0) :: field_1_w0
+    END SUBROUTINE setval_field_w0_code
+  END MODULE setval_field_w0_mod
+```
+
+The generated subroutine body with the argument list and declarations can
+be used to complete the kernel stub above (note that the generated
+`USE constants_mod` statement is redundant as per LFRic convention it is
+already declared at the module level).
+
+`rscalar_2` denotes the read-only scalar argument passed from the
+[LFRic algorithm layer](LFRic_algorithm.md) as is. As said
+[above](#what-kernels-do), kernels operate on a subset of a field
+object and the generated code reflects the required information for a
+kernel to read from and update a field:
+* `nlayers` is the number of vertical layers in the mesh that the
+  field's DoFs are placed on (or the number of cells in a vertical
+  column that the kernel operates on);
+* `field_1_w0` is the data stored in an LFRic field object;
+* `ndf_w0` is the number of all DoFs (owned and annexed) for the field
+  argument defined on `W0` function space;
+* `undf_w0` is the number of owned DoFs for the field argument defined
+  on `W0` function space;
+* `map_w0` is the DoF-map for the cell at the base of the column for
+  the `W0` function space that this field is defined on (the map
+  contains local IDs of DoFs in a partition).
+
+#### Loops
+
+Having defined the kernel metadata, subroutine arguments and
+declarations, we can write the executable code to actually update one
+cell-column of a field in this example. In LFRic kernels this is done by
+an outer `k`-loop over layers that updates each cell in a column and an
+inner loop over DoFs that updates the DoFs in a cell.
+
+```fortran
+do k = 0, nlayers-1
+  do df = 1, ndf_w0
+    field_1_w0( map_w0(df) + k ) = rscalar_2
+  end do
+end do
+```
```

Comment:
Looping over k in the kernel is an optimisation which assumes that the data is k-first. In the future we may move this looping up to the PSy layer to better handle i-first fields. Should this be noted or omitted as we are still quite far from actually implementing that.

Reply:
This was also raised by PSyclons :) Added a note to clarify.